source/common/router/router.cc

```diff
@@ -219,8 +220,23 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
       return Http::FilterHeadersStatus::StopIteration;
     }
     config_.stats_.rq_direct_response_.inc();
-    sendLocalReply(route_->directResponseEntry()->responseCode(), "", false);
-    // TODO(brian-pane) support sending a response body and response_headers_to_add.
+    std::string body;
+    Optional<std::string> inline_body = route_->directResponseEntry()->responseBody();
+    if (inline_body.valid()) {
+      body = inline_body.value();
+    }
+    Optional<std::string> path = route_->directResponseEntry()->responseBodyFilename();
+    if (path.valid()) {
+      const ssize_t kMaxFileSize = 4096;
+      ssize_t file_size = Filesystem::fileSize(path.value());
+      if (file_size < 0 || file_size > kMaxFileSize) {
+        sendLocalReply(Http::Code::InternalServerError, "", false);
+        return Http::FilterHeadersStatus::StopIteration;
+      }
+      body = Filesystem::fileReadToEnd(path.value());
```

Comment:
This may throw if path is a directory or exists but is not readable. I expect that probably turns into an internal server error elsewhere, but wanted to point it out in case that's not the desired behavior.

Reply:
I would strongly advise not reading this file during serving. Instead, I would read it into memory during route config load, and then just sent the pre-read buffer. Length sanity checking, etc. can be handled at that point.

Reply:
@mattklein123 are you proposing to watch for changes to the file (e.g., by adding an inotify fd to an existing epoll loop) and re-read when it changes?

Reply:
I'm proposing just reading it once at config load time, and documenting very well the behavior. In general, I feel that this should be generally sufficient for the ways that people are likely to use this behavior. (More generally, I would prefer for Envoy to not become a generic web server, and if we go in that direction I feel that it should be done in a new filter and not as part of the router).

I think this is much easier to reason about from an error handling perspective.

Reply:
My concern is whether this would set a bad precedent by being the first Envoy feature that requires a manual restart or reload to pick up a configuration change. Or do we already have that issue with TLS cert and key files?

Reply:
There are ways to workaround this, for example to rename the file as part of operations, etc. IMO this is an OK compromise for this feature as long as it is well documented. We could add watching in the future as a separate feature if desired.

If we want to move forward w/ this implementation we definitely must catch exceptions and handle, as well as understand potential perf issues (though the file will likely be in cache).

Reply:
Sounds good, I'll go with the simple read-at-startup approach for now.

Reply:
I think read-at-startup is fine for this feature, brian you can ignore the rest of this :-P

That said, do we have a plan for files for v2 in general?  This could be enabling pushing files via RPC, some config command which says "reload files and tell me if config is valid", disallowing files when you're getting live config or even documenting "rename rather than change contents of your files or you will shoot yourself in the foot".   @htuch

Reply:
@alyssawilk in general I think we do not currently have a great plan/roadmap for files. This is exemplified in https://github.com/envoyproxy/envoy/issues/2241. IMO if we expect Envoy to really start heavily using files in substantial production deployments we need to put a lot more thought into the entire situation (e.g., there are definitely bugs in the current code where if someone deletes an xDS file after initial error checking Envoy will probably crash). We should probably track in a dedicated ticket.

Reply:
Off the top of my head, I think an RPC to push file changes to Envoy would be a good thing. It would solve all the atomicity problems associated with people using cp rather than mv. In addition, if there were a RESTful push interface, it could provide a simpler alternative to the long-poll JSON-REST API.

And it would be nice to have a way to push TLS certificates and keys to Envoy without having to store them in a filesystem on the local host.